jax/_src/lax/control_flow/loops.py

```diff
@@ -263,106 +286,197 @@ def _create_jaxpr(init):
 
   _check_tree_and_avals("scan carry output and input",
                         # Extract the subtree and avals for the first element of the return tuple
-                        out_tree_children[0], carry_avals_out,
+                        out_tree_children[1], carry_avals_out,
                         init_tree, carry_avals)
   disallowed_effects = jaxpr.effects - allowed_effects
   if disallowed_effects:
     raise NotImplementedError(
         f'Effects not supported in `scan`: {disallowed_effects}')
 
-  out = scan_p.bind(*consts, *in_flat,
+  # use default precision with no weak types
+  start = jax.numpy.array([0])[0]
+  stop = jax.numpy.array([length])[0]
+  done = jax.numpy.array(early_exit(init), dtype=np.bool_)
+  out = scan_p.bind(start, stop, done, *consts, *in_flat,
                     reverse=reverse, length=length, jaxpr=jaxpr,
                     num_consts=len(consts), num_carry=len(init_flat),
-                    linear=(False,) * (len(consts) + len(in_flat)),
+                    linear=(False,) * (3 + len(consts) + len(in_flat)),
                     unroll=unroll)
-  return tree_unflatten(out_tree, out)
-
-def _scan_impl_unrolled(*args, reverse, length, num_consts, num_carry, linear,
-                        f_impl, x_avals, y_avals):
-  consts, init, xs = split_list(args, [num_consts, num_carry])
+  out = out[2:]  # trim off start and stop
+  _, carry, ys = tree_unflatten(out_tree, out)  # trim off done
+  return carry, ys
```

Comment:
The primitive returns `start`, `stop`, and `done` for use as residuals in the rules below. (And so they need to be trimmed off here before returning to the user.)